drivers/crypto/crypto_mchp_sha_g1.c

```diff
@@ -0,0 +1,348 @@
+			    const unsigned char *data, unsigned int len, unsigned char *digest)
+{
+	int block_size = algo_cfg->block_size;
+	unsigned int processed = 0;
+	int ret;
+
+	regs->SHA_CR = SHA_CR_SWRST_Msk;
+	regs->SHA_MR = algo_cfg->sha_mr_algo | SHA_MR_SMOD_AUTO_START | SHA_MR_PROCDLY_LONGEST;
+	regs->SHA_CR = SHA_CR_FIRST_Msk;
+	regs->SHA_MSR = len;
+	regs->SHA_BCR = len;
+
+	if (len == 0) {
+		/* For the empty message, automatic padding is not quired in this driver
+		 * (SHA_MSR.MSGSIZE and SHA_BCR.BYTCNT are configured to 0). The block to
+		 * be processed are the padded part (a one bit, then  '1' followed by zeor
+		 * bits) and then the message length.
+		 */
+		regs->SHA_IDATAR[0] = FIRST_WORD_4_PADDING_EMPTY_MSG;
+		for (int i = 1; i < block_size / sizeof(uint32_t); i++) {
+			regs->SHA_IDATAR[i] = 0;
+		}
+		goto out;
+	}
+
+	while (len - processed >= block_size) {
+		/* Write the block to be processed in the Input Data Registers */
+		mchp_sha_set_input(regs, &data[processed], block_size);
+		ret = mchp_sha_wait_data_rdy(regs);
+		if (ret < 0) {
+			return ret;
+		}
+
+		processed += block_size;
+	}
+
+	if (len - processed) {
+		mchp_sha_set_input(regs, &data[processed], len - processed);
+		ret = mchp_sha_wait_data_rdy(regs);
+		if (ret < 0) {
+			return ret;
+		}
+
+		processed += block_size;
+	}
+
+out:
+	ret = mchp_sha_wait_data_rdy(regs);
+	if (ret < 0) {
+		return ret;
+	}
+	mchp_sha_get_output(regs, digest, algo_cfg->dgst_len);
+
+	return ret;
+}
+
+static struct crypto_mchp_sha_session *crypto_mchp_sha_get_unused_session(void)
+{
+	k_sem_take(&mchp_sha_session_sem, K_FOREVER);
+
+	for (int i = 0; i < ARRAY_SIZE(mchp_sha_sessions); i++) {
+		if (!mchp_sha_sessions[i].in_use) {
+			mchp_sha_sessions[i].in_use = true;
+			k_sem_give(&mchp_sha_session_sem);
+
+			return &mchp_sha_sessions[i];
+		}
+	}
+
+	k_sem_give(&mchp_sha_session_sem);
+
+	return NULL;
+}
+
+static int mchp_sha_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	const struct device *dev = ctx->device;
+	const struct crypto_mchp_sha_cfg *cfg = dev->config;
+	struct crypto_mchp_sha_data *data = dev->data;
+	struct crypto_mchp_sha_session *session = ctx->drv_sessn_state;
+	bool inplace_ops = (ctx->flags & CAP_INPLACE_OPS) == CAP_INPLACE_OPS;
+	int ret;
+
+	if (!pkt || !pkt->in_buf || (!pkt->out_buf && !inplace_ops)) {
+		LOG_ERR("Invalid packet buffers");
+		return -EINVAL;
+	}
+
+	if (!finish) {
+		LOG_ERR("Multipart shaing not supported yet");
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	if (session->algo_cfg == NULL) {
+		k_sem_give(&data->device_sem);
+		LOG_ERR("Unsupported algorithm");
+		return -ENOTSUP;
+	}
+
+	if (inplace_ops && pkt->in_len < session->algo_cfg->dgst_len) {
+		LOG_ERR("Insufficient in_buf for digest");
+		return -EINVAL;
+	}
+
+	ret = mchp_sha_process(cfg->regs, session->algo_cfg, pkt->in_buf, pkt->in_len,
+			       inplace_ops ? (uint8_t *)(uintptr_t)pkt->in_buf : pkt->out_buf);
```

Comment:
why are you casting `in_buf` to `uintptr_t` to (uint8_t *) ? It does not seem necessary this cast to `uintptr_t`

Reply:
couldn't you simply `(uint8_t *)pkt-in_buf` ?